storage: watchable_store.go clean up

gyuho opened the pull request:
1. change unsynced to `map[string]map[*watcher]struct{}`
(coreos#4136, kills TODO)

2. separate the function that gets all events to `eventsFromKVs`
(based on Anthony's feedback)

3. wrap map[string]map[*watcher]struct{} as watcherSetByKey

gyuho commented:
/cc @xiang90 @heyitsanthony Please take a look. Thanks,

heyitsanthony commented on storage/watchable_store.go:
```diff
@@ -448,13 +381,99 @@ type watcher struct {
+// kvsFromUnsynced returns all key bytes and raw storagepb.KeyValue bytes
+// for all unsynced watchers. And returns prefixes, a map of prefixes that
+// need to be watched. store must be mutex locked before this function.
+func kvsFromUnsynced(s *watchableStore) ([][]byte, [][]byte, map[string]struct{}) {
```
The returned 3-tuple would be easier to work with here and elsewhere as a struct (`PrefixedKVs`? `WatchKVs`?)

gyuho replied:
Ok I will wrap them as `WatchKVs` struct. Thanks,

heyitsanthony commented on storage/watchable_store.go:
```diff
@@ -448,13 +381,99 @@ type watcher struct {
+// raw storagepb.KeyValue bytes. It passes pointers of map in case
+// the map contains large amount of data. The maps aren't protected.
+// Make sure to lock for thread-safety.
+func eventsFromKVs(keys [][]byte, vs [][]byte, prefixes *map[string]struct{}, sm *map[string]map[*watcher]struct{}) []storagepb.Event {
```
this could be a method on `WatchKVs` (or whatever it'll be called)

gyuho replied:
Ok

heyitsanthony commented on storage/watchable_store.go:
```diff
@@ -46,7 +46,7 @@ type watchableStore struct {
 
 	// contains all unsynced watchers that needs to sync with events that have happened
-	unsynced map[*watcher]struct{}
+	unsynced map[string]map[*watcher]struct{}
```
Why not `map[string]*watcher`?

heyitsanthony commented on storage/watchable_store.go:
```diff
@@ -64,7 +64,7 @@ type cancelFunc func()
 	s := &watchableStore{
 		store:    NewStore(b, le),
-		unsynced: make(map[*watcher]struct{}),
+		unsynced: make(map[string]map[*watcher]struct{}),
```
this is a complicated type-- `type WatchersByKey map[string]map[*watcher]struct{}`? `KeyedWatchers`? Not sure on a good name

gyuho replied:
Good idea. Will use `WatchersByKey`

gyuho commented:
@heyitsanthony All addressed. PTAL. Thanks!

xiang90 commented on storage/watchable_store.go:
```diff
@@ -40,17 +40,19 @@ type watchable interface {
 	rev() int64
 }
 
+type watchersByKey map[string]map[*watcher]struct{}
```
let define ?
```go
watcherSet map[*watcher]struct{}

watcherSetByKey map[string]watcherSet
```

/cc @heyitsanthony

gyuho replied:
Ok will change.

xiang90 commented on storage/watchable_store_test.go:
```diff
@@ -215,6 +216,88 @@ func TestSyncWatchers(t *testing.T) {
 	}
 }
 
+func Test_kvsFromUnsynced(t *testing.T) {
```
remove the underscore?

xiang90 commented on storage/watchable_store.go:
```diff
@@ -448,13 +383,105 @@ type watcher struct {
+	// values are actual key-value pairs in backend.
+	tx := s.store.b.BatchTx()
+	tx.Lock()
+	wv.keys, wv.vs = tx.UnsafeRange(keyBucketName, minBytes, maxBytes, 0)
```
we cannot do this. We have to copy the bytes returned from tx before closing the tx if you want to use it later. I would suggest to keep the old logic to use the kv in place without the copying.

xiang90 commented on storage/watchable_store.go:
```diff
@@ -448,13 +383,105 @@ type watcher struct {
+}
+
+// kvsFromUnsynced returns watchKVs. store must be mutex locked before this function.
+func kvsFromUnsynced(s *watchableStore) watchKVs {
```
unsafeGetKVsToSync

gyuho commented:
Just did second clean up. We wrap the map with custom type. So I just converted
some functions with long names with methods. Now it should read more straightforward.
And removed `unsafeGetKVsToSync` (or `unsafeGetKVsToSync`) and just leave it inside
`syncWatchers` for this PR. 

@xiang90  @heyitsanthony  PTAL

xiang90 commented:
@gyuho @heyitsanthony 

I still feel the pull request is messy. I am going to try to rework it later this week. Then I probably can have more insight.

gyuho commented:
Sure. Thanks!
